src/include/test/unit_tdb_matrix_with_ids.cc

```diff
@@ -0,0 +1,180 @@
+  int offset = 13;
+  size_t Mrows = 200;
+  size_t Ncols = 500;
+  auto X = ColMajorMatrixWithIds<TestType, TestType, size_t>(Mrows, Ncols);
+  fill_and_write_matrix(
+      ctx, X, tmp_matrix_uri, tmp_ids_uri, Mrows, Ncols, offset);
+  CHECK(X.ids()[0] == offset + 0);
+  CHECK(X.ids()[1] == offset + 1);
+  CHECK(X.ids()[10] == offset + 10);
+
+  auto Y = tdbColMajorMatrixWithIds<TestType, TestType>(
+      ctx, tmp_matrix_uri, tmp_ids_uri);
+  Y.load();
+  CHECK(num_vectors(Y) == num_vectors(X));
+  CHECK(dimension(Y) == dimension(X));
+  CHECK(
+      std::equal(X.data(), X.data() + dimension(X) * num_vectors(X), Y.data()));
+  for (size_t i = 0; i < X.num_rows(); ++i) {
+    for (size_t j = 0; j < X.num_cols(); ++j) {
+      CHECK(X(i, j) == Y(i, j));
+    }
+  }
+  CHECK(X.num_ids() == Y.num_ids());
+  CHECK(std::equal(X.ids().begin(), X.ids().end(), Y.ids().begin()));
+  for (size_t i = 0; i < X.num_ids(); ++i) {
+    CHECK(X.ids()[i] == Y.ids()[i]);
+  }
+
+  auto Z = tdbColMajorMatrixWithIds<TestType, TestType>(std::move(Y));
+  CHECK(num_vectors(Z) == num_vectors(X));
+  CHECK(dimension(Z) == dimension(X));
+  CHECK(
+      std::equal(X.data(), X.data() + dimension(X) * num_vectors(X), Z.data()));
+  for (size_t i = 0; i < 5; ++i) {
+    for (size_t j = 0; j < 5; ++j) {
+      CHECK(X(i, j) == Z(i, j));
+    }
+  }
+
+  CHECK(X.num_ids() == Z.num_ids());
+  CHECK(std::equal(X.ids().begin(), X.ids().end(), Z.ids().begin()));
+  for (size_t i = 0; i < X.num_ids(); ++i) {
+    CHECK(X.ids()[i] == Z.ids()[i]);
+  }
+}
+
+TEMPLATE_TEST_CASE(
+    "tdb_matrix: assign to matrix", "[tdb_matrix_with_ids]", float, uint8_t) {
+  tiledb::Context ctx;
+  std::string tmp_matrix_uri = "/tmp/tmp_tdb_matrix";
+  std::string tmp_ids_uri = "/tmp/tmp_tdb_ids_matrix";
+  int offset = 123;
+  size_t Mrows = 200;
+  size_t Ncols = 500;
+
+  auto X = ColMajorMatrixWithIds<TestType, TestType, size_t>(Mrows, Ncols);
+  fill_and_write_matrix(
+      ctx, X, tmp_matrix_uri, tmp_ids_uri, Mrows, Ncols, offset);
+  CHECK(X.ids()[0] == offset + 0);
+  CHECK(X.ids()[1] == offset + 1);
+  CHECK(X.ids()[10] == offset + 10);
+
+  auto B = ColMajorMatrixWithIds<TestType, TestType, size_t>(0, 0);
+  {
+    auto Y = tdbColMajorMatrixWithIds<TestType, TestType>(
+        ctx, tmp_matrix_uri, tmp_ids_uri);
+    Y.load();
+    B = std::move(Y);
+  }
+
+  {
+    auto Y = tdbColMajorMatrixWithIds<TestType, TestType>(
+        tdbColMajorMatrixWithIds<TestType, TestType>(
+            ctx, tmp_matrix_uri, tmp_ids_uri));
+  }
+
+  auto Y = tdbColMajorMatrixWithIds<TestType, TestType>(
+      ctx, tmp_matrix_uri, tmp_ids_uri);
+  Y.load();
+
+  CHECK(num_vectors(Y) == num_vectors(X));
+  CHECK(dimension(Y) == dimension(X));
+  CHECK(
+      std::equal(X.data(), X.data() + dimension(X) * num_vectors(X), Y.data()));
+  for (size_t i = 0; i < 5; ++i) {
+    for (size_t j = 0; j < 5; ++j) {
+      CHECK(X(i, j) == Y(i, j));
+    }
+  }
+
+  // Check that we can assign to a matrix
```

Comment:
How should we define inter-operation between a `Matrix` and a `MatrixWithIds`?  Should they be able to assign to / copy each other?  We should define this behavior and add appropriate unit tests here and in `unit_matrix_with_ids.cc`

Reply:
So today it looks like we can move from a `MatrixWithIds` to a `Matrix` okay, but not the other way around. What is your preference here on end behavior? It seems like we won't be doing this move often so I would be fine preventing it (if it's possible which I'm guessing it is)?